Added runtime method to determine resolved package versions.

mitchell-as opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://activestatef.atlassian.net/browse/DX-2408" title="DX-2408" target="_blank"><img alt="Story" src="https://activestatef.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10308?size=medium" />DX-2408</a>  We have a runtime mechanic for resolving requested versions to actual versions
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->
<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

Demo in action: #3017 

We don't need to do any normalization at the moment because any inputs would be coming from Platform API calls (e.g. in the demo above, names come from requirements). If we want to do any normalization (e.g. from user input), we already have inventory.go's `model.FetchNormalizedName()`, which can be used in conjunction with this function if need be.

mitchell-as commented:
I'm not 100% sure this satisfies the ACs, but I'm putting it up for review for feedback on what I'm missing.

Naatan reviewed (changes requested):
This is useful for the `state install` use-case bit for cases like `state packages` I think we should be able to more proactively provide a mapping of packages and their resolved versions. Two things I'd like to see this PR include:

1. The runtime mechanic should already know which packages were requested. So from an external point of view I should just be able to call something like `ResolvedVersions()` which returns a map of all the normalized package names and their resolved version.
  * I realize normalization might be a bit too foreign to the runtime mechanic, so possibly we can drop that AC and handle that at a different level.
2. We should not be re-initializing the buildplan for each resolution. I think we already parse it whenever the runtime is read? If not we can probably add that logic somewhere (in an on-demand type of way).

mitchell-as commented:
Okay, I changed the method to be `ResolvedArtifacts()`, which among other things, contains the resolved version number for packages. The returned name is the normalized version, so callers are responsible for normalizing names prior to comparisons. I've updated #3017 to utilize the new method (https://github.com/ActiveState/cli/pull/3017/files#diff-67826c0e717c1b1a1a3072382df29bb908a524e39bc69d3742619b79b306c943R205-R211)

Naatan commented on pkg/platform/runtime/runtime.go:
```diff
@@ -356,3 +358,29 @@ func (r *Runtime) ExecutableDirs() (envdef.ExecutablePaths, error) {
+			runtimeStore = store.New(r.target.Dir())
+		}
+
+		plan, err := runtimeStore.BuildPlan()
+		if err != nil {
+			return nil, errs.Wrap(err, "Unable to fetch build plan")
+		}
```
This would still mean the buildplan gets parsed twice if it was used for other purposes (eg. runtime was updated). Given that this is a costly operation we may want to introduce some caching here too? Suggest checking with Mike, he may have some ideas/opinions.

This is based on my assumption that parsing buildplans still has a performance impact that is noticeable by users. If I'm wrong on that I'll drop this concern.

mitchell-as replied:
After a discussion on Slack, Mike said "As far as calling this method goes I don’t think it’s a big performance issue. This is returning the build structure which is created by parsing the build plan, so we aren’t parsing again to get it."

The function is just unmarshaling json into a Go struct representing an already parsed build plan.
